useeffect-and-calling-apis content (#45)
Ada-Developers-Academy/core-unit-3: useeffect-and-calling-apis/useeffect-after-render.md

| --------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `useEffect(..., ...);`                              | We still call `useEffect`                                                                                                                                                       |
| `() => { ... }`                                     | Our first argument is still a function that should execute after the component is rendered                                                                                      |
| `[]`                                                | The second argument is an empty array, to express that there _are no dependencies_ to watch for. Essentially, there are _no_ dependencies that should cause `useEffect` to run. |

## Example: Changing `document`

Let's explore an example React app that will showcase the importance of when `useEffect` runs.

Just like static-site development using HTML, CSS, and vanilla JS, JavaScript can execute faster than the DOM is built.

> In vanilla JS, we handled this problem by listening for the event "DOMContentLoaded":

Author: Thanks! I didn't actually know what was happening/I had always thought it was a race condition, so I appreciate the TIL :D

### !challenge
* type: checkbox
* id: 4e323978
* title: useEffect After Render
##### !question

Consider this code:

```js
import { useState, useEffect } from 'react';

const Timer = () => {
  const [currentTime, setCurrentTime] = useState(0);
  const [isEnabled, setIsEnabled] = useState(false);

  useEffect(() => {

  }, []);
}

```

`useEffect` will run every time certain value(s) update. Which values will cause `useEffect` to run?

##### !end-question
##### !options

* `currentTime`
* `isEnabled`
* `Timer`
* None of the above

##### !end-options
##### !answer

* None of the above

##### !end-answer
### !end-challenge
<!-- prettier-ignore-end -->

### !callout-danger

## Our Examples Don't Utilize a Cleanup Function

On the subject of `useEffect` syntax, the `useEffect` hook also supports defining a _clean-up_ function. "Clean-up functions" describe functions responsible for "cleaning up potential side-effects," such as closing database connections.

<br/>

In the examples in this curriculum, there will not be many or any examples that use the cleanup function meaningfully.

<details>

<summary>For those curious, here's an example of defining a clean-up function.</summary>

> 

Reviewer: spacing (we're in a callout)
```suggestion

<br/>

```

In React, clean-up functions will run during `componentWillUnmount`. To define a clean-up function, we must _return_ the clean-up function inside our `useEffect` callback function.

<!-- prettier-ignore-start -->
```js

> import { useEffect } from 'react';
> 
> function App() {
> 
>   useEffect(() => {
>     console.log('I\'m in useEffect!');
>     console.log('This will be called whenever an instance of this component mounts');
> 
>     return () => { console.log('useEffect calls the returned function during componentWillUnmount') };
>   }, []);
> 
>   return (<h1>Demo of Clean-up Function Syntax</h1>);
> }

Reviewer: the current example doesn't really show the unmount behavior occurring, so I fleshed this out a bit
```suggestion
import { useState, useEffect } from 'react';

const ComponentWithCleanup = () => {

  useEffect(() => {
    console.log('I\'m in useEffect!');
    console.log('This will be called whenever an instance of this component mounts');

    return () => { console.log('useEffect calls the returned function during componentWillUnmount') };
  }, []);

  return (<h1>Demo of Clean-up Function Syntax</h1>);
};

function App() {
  const [showComponent, setShowComponent] = useState(true);

  const childComponent = showComponent ? <ComponentWithCleanup /> : <h1>Removed!</h1>;

  return (
    <div>
      { childComponent }
      <button onClick={() => setShowComponent(! showComponent)}>Update</button>
    </div>
  );
}
```

```
<!-- prettier-ignore-end -->

> In this example, our `useEffect` callback function returns a one-line, anonymous arrow function, which prints to the console.

Reviewer: expand explanation to match revised example
```suggestion
<br/>

In this example we define a `ComponentWithCleanup` component, which logs a message to the console in its cleanup function. The `App` component has a button that toggles whether to include the `ComponentWithCleanup` in the rendered output.

<br/>

When the app is rendered, we see the messages logged to the console through the component's `useEffect` function. Then when we click the button, the component is removed from the rendered output, and we see the cleanup message appear in the console log.
```
